File: books/cs/introduction-to-algorithms/chapter6/C/heap_sort.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <math.h>

#define DATASETSIZE 10

void heap_sort(double* srcdata, int len);
void main(void){
    double srcdata[DATASETSIZE];
    int idx, jdx;

    srand(time(NULL));
    for( idx=0; idx<DATASETSIZE; idx++ ){
        srcdata[idx] = rand()*1.0/RAND_MAX;
    }
    for( idx=0; idx<DATASETSIZE; idx++ ){
        if(idx%10 == 0) printf("\n\n");
        printf(" %f, ",srcdata[idx]);
    }
    printf("\n");
    heap_sort(srcdata,DATASETSIZE);
    for( idx=0; idx<DATASETSIZE; idx++ ){
        if(idx%10 == 0) printf("\n\n");
        printf(" %f, ",srcdata[idx]);
    }
    printf("\n");
}

void swap(double* srcdata,int i,int j){
    double temp;

    temp = srcdata[i];
    srcdata[i] = srcdata[j];
    srcdata[j] = temp;
}

int left(int i){
        return 2*i;
}

int right(int i){
        return 2*i+1;
}


void max_heapify(double* srcdata,int len,int i){
        int l = left(i);
        int r = right(i);
        int largest = 0;

        if( l<len && srcdata[l]>srcdata[i] ){
                largest = l;
        }else largest = i;
        if( r<len && srcdata[largest]<srcdata[r] ){
                largest = r;
        }
        if( largest!=i ){
                swap(srcdata,largest,i);
                max_heapify(srcdata,len,largest);
        }
}

void build_max_heap(double* srcdata,int len){
        int idx=0;

        for(idx=(int)(floor(len/2));idx>=0;idx--){
                max_heapify(srcdata,len,idx);
        }
}

void heap_sort(double* srcdata,int len){
        int idx=0;

        build_max_heap(srcdata,len);
        for(idx=len-1;idx>0;idx--){
                swap(srcdata,0,idx);
                len--;
                max_heapify(srcdata,len,0);
        }
}

File: books/cs/introduction-to-algorithms/chapter6/R/heap_sort.R
#
# heap sort algorithm
# introduction to algorithm, 3rd edition, chapter 6
#
# clean the workspace and memory

rm( list=ls() )
gc()

aaa <- sample(999,10,replace = T)
srcdata <- aaa[aaa>100]

swap <- function(i,j){
  temp <- srcdata[i]
  srcdata[i] <<- srcdata[j]
  srcdata[j] <<- temp
}

partition <- function(p,r){
  print(c(p,r,"",length(srcdata),srcdata))
  x <- srcdata[r]
  idx <- p-1
  for(jdx in p:(r-1)){
    if( srcdata[jdx] >= x ){
      idx <- idx + 1
      swap(idx,jdx)
    }
  }
  swap(idx+1,r)
  print(c(p,r,idx+1,length(srcdata),srcdata))
  return( idx+1 )
}

quick_sort <- function(p,r){
  if( p<r ){
    q <- partition(p,r)
    quick_sort(p,q-1)
    quick_sort(q+1,r)
    print(c(p,r, srcdata))
  }
}

left <- function(i){
  return(2*i)
}

right <- function(i){
  return(2*i+1)
}

max_heapify <- function(len,i){
  l <- left(i)
  r <- right(i)
  largest <- 0
  
  if( l<=len && srcdata[l]>srcdata[i] ){
    largest <- l
  }else largest <- i
  if( r<=len && srcdata[r]>srcdata[i] ){
    largest <- r
  }
  
  if( largest != i ){
    swap(i,largest)
    max_heapify(len,largest)
  }
}

build_max_heap <- function(len){
  
  for( idx in floor(len/2):1 ){
	max_heapify(len,idx)
  }
}

heap_sort <- function(len){
  
  build_max_heap(len)
  for( idx in len:2 ){
	swap(idx,1)
	len <- len-1
	max_heapify(len,1)
  }
}

print(srcdata)
#quick_sort(1,length(srcdata))
heap_sort(length(srcdata))
print(srcdata)
